WIP Buying and selling concept 1

ultidonki opened the pull request:
Please don't merge! 

This PR contains all of the functionality of buying and selling that I want to add. And it should be bug-free. 
So feel free to play with it, or code-review it. 

But there's all sorts of weird stuff I want to find a cleaner solution for. (For example, I might split the ShopMenuState into both a BuyShopMenuState, and a SellShopMenuState, because right now I only know whether the player is buying or selling by checking if buyer == None. Which is a bit hacky, but it's how it was working before I started looking at it.)

I might have more time to clean it up in the new year. (Unless someone else wants to clean it up first!)

Here's what it looks like: 
![buying_shop_2](https://user-images.githubusercontent.com/89098769/145722046-a9d2dffd-daa1-4802-af8b-0030a164889d.png)

Features include:
- You can now create 'economies' in .json, which tell you the sell/buy price of a certain shop. See #429 
- Economies have parent-child relationships, so if a buy/sell price isn't in the current economy, it'll look up the buy/sell price in the parent economy.
- Player now has a game variable of 'money', used as currency in the shop. See #668 
- Buying now shows you: Price of item, how much money the buyer has, how many items in the bag the buyer already has. 
- Selling now shows you: Price you'll gain by selling, how much money the seller has, how many items in the bag the seller already has. 
- Buying and selling now increases/decreases the 'money' in both the buyer and seller's 'money' variable. 
- When buying, you cannot choose a quantity higher than you can afford. 

There's a temp map for testing in there - shops.tmx 
On this map, you gain 1 potion per frame, so the amount of potions you have in your bag will always be wrong, as it increases even when you're on the buying/selling menu. 
If you take 1 step in any direction, you'll gain $100. 
(Which should make testing it easier)

If you sell the item at the bottom of the sell menu, the game crashes with an 'assert selected' error. This happens in development too, I haven't had time to fix it.

Commit:
Show price in shop buy state to match quantity

Commit:
Show player's current money while in buy shop

Commit:
Add initial economy jsons

Commit:
Add test shopping map and give money if moved event

Commit:
Don't show buyer's amount of money when on Sell screen

Commit:
Add level 999 propellercat to dungeon crawl encounters

Commit:
Load economy from JSON, add action to set it in a shop

Commit:
Make items lookup their prices in an economy during buying

Commit:
Get recursive economy item price lookup working

Commit:
Buyers and sellers now spend and gain money in Buy menu

Commit:
Stop buyer entering quantity higher than they can afford

Commit:
Show qty already in buyer's bag when buying

Commit:
Fix sell menu, and show correct variables when in the sell menu

bitcraft commented on tuxemon/item/economy.py:
```diff
@@ -0,0 +1,500 @@
+        # Auto-load the economy from the economy database.
+        self.load(slug)
+
+    # class Item:
```
in general, please don't leave in commented code.  there are cases where we could make exceptions, but this is just commented out, and in a few months, nobody will know why or what it does.

bitcraft commented on mods/tuxemon/db/economy/dungeon_crawl_ultidonki.json:
```diff
@@ -0,0 +1,24 @@
+{
```
minor: if your ide or editor supports it, enabling formatting makes the json easier to follow.  you don't need to change it though.

bitcraft commented on mods/tuxemon/maps/shops.tmx:
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<map version="1.2" tiledversion="1.3.2" orientation="orthogonal" renderorder="right-down" compressionlevel="0" width="8" height="8" tilewidth="16" tileheight="16" infinite="0" nextlayerid="5" nextobjectid="7">
+ <layer id="1" name="Tile Layer 1" width="8" height="8">
+  <data encoding="csv">
```
in the future, please make sure the maps use the "base64 gzipped" format.  it produces faster loading and smaller maps.

bitcraft commented on tuxemon/event/actions/set_economy.py:
```diff
@@ -0,0 +1,62 @@
+        npc = get_npc(self.session, self.parameters.npc_slug)
+        assert npc
+        if self.parameters.economy_slug is None:
+            npc.economy = Economy("default")
```
please put "If the economy name is not specified, the economy named "default" will be used" in the docstring

bitcraft commented on tuxemon/item/economy.py:
the search function for finding items by price should be generalized and shared for use by any lookup methods

bitcraft commented on tuxemon/npc.py:
```diff
@@ -141,25 +162,36 @@ def __init__(
 
         # combat related
-        self.ai: Optional[AI] = None  # Whether or not this player has AI associated with it
+        self.ai: Optional[
```
i can tell that you used black on this file.  not a problem, but lines like this should be modified by moving the inline comment to the line above it so that it doesn't leave this imo bad looking format which preserves the inline comment.

bitcraft commented on tuxemon/npc.py:
```diff
@@ -779,5 +883,52 @@ def alter_item_quantity(
 
         return success
 
+    def alter_item_quantity_and_money(
```
this function is really taking on a lot of behavior.  it would be nice to see this split up into smaller parts so it can be easily tested and would make maintenance easier.

bitcraft commented on tuxemon/states/items/__init__.py:
```diff
@@ -292,34 +298,57 @@ def on_menu_selection(self, menu_item: MenuItem[Item]) -> None:
             None if item_dict.get("infinite")
             else item_dict["quantity"]
         )
+        qty_in_bag = (
```
could you put a comment over this?  its not immediately obvious what this is doing.

bitcraft reviewed (commented):
left a few thoughts on this for now.  its a lot of changes, and still WIP so you can treat it as just comments for now.  i will need to spend some time testing it in game.  thank you for including a test map, that will make testing much easier.

ultidonki commented:
Oh, yes - I wasn't intending to add any commented-out lines or run the black formatter! :scream: 

The rest of the comments all sound fine - I'll try to address them when I have time.

ultidonki commented:
I'm going to split this pull request into several smaller ones, over the next couple of months. 
I'm struggling to understand my own code in places - there's a lot of 'nice-to-have' code jumbled in with the necessary code to get shops to work, making the pull request really hard for me to understand myself! So I'm sure everyone will struggle if they try to read it. 
I'll leave this here to make sure I address each of the comments above when I tackle each part of the fixed pull request, and then close this PR when all the smaller ones are closed.

ultidonki commented:
I'm going to close this PR now - 90% of the buying/selling code in this PR has gone into a different PR, #1144. 
I've addressed all of the comments above in the new PR, and I've cleaned up some other stuff - so Buying and Selling will become their own subclass in the new PR, which is much better. So I'll close this one.
